src/ATen/native/xpu/sycl/SoftMaxKernels.cpp

```diff
@@ -1712,6 +1714,221 @@ void spatial_softmax_backward(
 #undef SPATIAL_SOFTMAX_BACKWARD_IMPL
 }
 
+template <typename scalar_t, typename accscalar_t, bool LogSoftMax>
+Tensor& masked_softmax_forward(
+    Tensor& output,
+    Tensor& input,
+    int dim,
+    const Tensor mask) {
+  auto inner_size = input.stride(dim);
+  auto dim_size = input.size(dim);
+  auto outer_size = input.numel() / (inner_size * dim_size);
+
+  constexpr int float4_size = sizeof(float) * 4;
+  constexpr int max_vec_size = float4_size / sizeof(scalar_t);
+  constexpr int INNER_LOOP = max_vec_size * 2;
+
+  // decide vec_size: max_vec_size or 1
+  using vec_t = at::native::memory::aligned_vector<scalar_t, max_vec_size>;
+  constexpr int align_bytes = alignof(vec_t);
+  int input_start =
+      ((uint64_t)input.data_ptr()) % align_bytes / sizeof(scalar_t);
+  int output_start =
+      ((uint64_t)output.data_ptr()) % align_bytes / sizeof(scalar_t);
+
+  // decide indexing range: uint32_t (4GB) or uint64_t (>4GB)
+  bool can_use_32bit_index =
+      canUse32BitIndexMath(input) && canUse32BitIndexMath(output);
+
+  // decide SIMD: SIMD32 or SIMD16
+  auto* dev_prop =
+      at::xpu::getDeviceProperties(at::xpu::getDeviceIndexOfCurrentQueue());
+  auto sub_group_size = dev_prop->sub_group_sizes;
+  int SIMD = sub_group_size[1];
+  if (SIMD == SIMD32) {
+    if (dim_size < SIMD16 * INNER_LOOP)
+      SIMD = SIMD16;
+  }
+
+#define DISPATCH_MASK_SOFTMAX_FORWARD_IMPL(vec_size, SIMD, outer_loop) \
+  {                                                                    \
+    dispatch_softmax_forward_kernel<                                   \
+        INNER_LOOP,                                                    \
+        vec_size,                                                      \
+        SIMD,                                                          \
+        scalar_t,                                                      \
+        accscalar_t,                                                   \
+        uint32_t,                                                      \
+        LogSoftMax,                                                    \
+        outer_loop,                                                    \
+        true,                                                          \
+        decltype(input_calc)>(                                         \
+        input.data_ptr<scalar_t>(),                                    \
+        output.data_ptr<scalar_t>(),                                   \
+        dim_size,                                                      \
+        outer_size,                                                    \
+        mask.data_ptr<bool>(),                                         \
+        input_calc);                                                   \
+  }
+
+  int max_group_size = 512;
```

Comment:
Can we have a comment here, if hard code 512. BTW, Can we query it from sycl api